File: app/src/main/java/com/elementary/tasks/core/views/MonthView.java
package com.elementary.tasks.core.views;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.os.Handler;
import androidx.annotation.IntRange;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

import com.elementary.tasks.core.calendar.Events;
import com.elementary.tasks.core.utils.AssetsUtil;
import com.elementary.tasks.core.utils.LogUtil;
import com.elementary.tasks.core.utils.Prefs;
import com.elementary.tasks.core.utils.ThemeUtil;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import hirondelle.date4j.DateTime;

/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

public class MonthView extends View implements View.OnTouchListener {

    private static final String TAG = "MonthView";
    private static final int ROWS = 6;
    private static final int COLS = 7;
    private static final int GRID_R_C = 3;
    private static final long LONG_CLICK_TIME = 500;

    private int mYear;
    private int mMonth;
    private int currentYear;
    private int currentMonth;
    private int currentDay;
    private List<DateTime> mDateTimeList;
    private Map<DateTime, Events> eventsMap = new HashMap<>();

    private Context mContext;

    private Paint paint;
    private Paint circlePaint;
    private List<Rect> mCells;
    private Map<Rect, List<Rect>> circlesMap = new HashMap<>();
    private int mWidth;
    private int mHeight;
    private int mDefaultColor;
    private int mTodayColor;
    private int mTouchPosition = -1;
    private Rect mTouchRect;

    private Handler mLongClickHandler = new Handler();
    private OnDateClick mDateClick;
    private OnDateLongClick mDateLongClick;
    private Runnable mLongRunnable = new Runnable() {
        @Override
        public void run() {
            mLongClickHandler.removeCallbacks(mLongRunnable);
            if (mTouchRect != null && mDateLongClick != null) {
                mDateLongClick.onLongClick(mDateTimeList.get(mTouchPosition));
            }
            cancelTouch();
        }
    };

    public MonthView(Context context) {
        super(context);
        init(context);
    }

    public MonthView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    public MonthView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context);
    }

    private void init(Context context) {
        this.mContext = context;
        this.paint = new Paint();
        this.paint.setAntiAlias(true);
        this.circlePaint = new Paint();
        this.circlePaint.setAntiAlias(true);
        ThemeUtil themeUtil = ThemeUtil.getInstance(context);
        if (themeUtil.isDark()) {
            mDefaultColor = Color.WHITE;
        } else {
            mDefaultColor = Color.BLACK;
        }
        mTodayColor = themeUtil.getColor(themeUtil.colorPrimary(Prefs.getInstance(context).getTodayColor()));
        paint.setTypeface(AssetsUtil.getTypeface(context, 7));
        Calendar calendar = Calendar.getInstance();
        calendar.setTimeInMillis(System.currentTimeMillis());
        currentDay = calendar.get(Calendar.DAY_OF_MONTH);
        currentMonth = calendar.get(Calendar.MONTH) + 1;
        currentYear = calendar.get(Calendar.YEAR);
        setDate(calendar.get(Calendar.YEAR), calendar.get(Calendar.MONTH) + 1);
        setOnTouchListener(this);
    }

    public void setDateClick(OnDateClick dateClick) {
        this.mDateClick = dateClick;
    }

    public void setDateLongClick(OnDateLongClick dateLongClick) {
        this.mDateLongClick = dateLongClick;
    }

    public void setEventsMap(Map<DateTime, Events> eventsMap) {
        this.eventsMap = eventsMap;
    }

    public void setDate(int year, @IntRange(from = 1, to = 12) int month) {
        mDateTimeList = new ArrayList<>();
        mDateTimeList.clear();
        mMonth = month;
        mYear = year;
        DateTime firstDateOfMonth = new DateTime(mYear, mMonth, 1, 0, 0, 0, 0);
        DateTime lastDateOfMonth = firstDateOfMonth.plusDays(firstDateOfMonth.getNumDaysInMonth() - 1);
        int weekdayOfFirstDate = firstDateOfMonth.getWeekDay();
        int startDayOfWeek = Prefs.getInstance(mContext).getStartDay() + 1;
        if (weekdayOfFirstDate < startDayOfWeek) {
            weekdayOfFirstDate += 7;
        }
        while (weekdayOfFirstDate > 0) {
            DateTime dateTime = firstDateOfMonth.minusDays(weekdayOfFirstDate
                    - startDayOfWeek);
            if (!dateTime.lt(firstDateOfMonth)) {
                break;
            }
            mDateTimeList.add(dateTime);
            weekdayOfFirstDate--;
        }
        for (int i = 0; i < lastDateOfMonth.getDay(); i++) {
            mDateTimeList.add(firstDateOfMonth.plusDays(i));
        }
        int endDayOfWeek = startDayOfWeek - 1;
        if (endDayOfWeek == 0) {
            endDayOfWeek = 7;
        }
        if (lastDateOfMonth.getWeekDay() != endDayOfWeek) {
            int i = 1;
            while (true) {
                DateTime nextDay = lastDateOfMonth.plusDays(i);
                mDateTimeList.add(nextDay);
                i++;
                if (nextDay.getWeekDay() == endDayOfWeek) {
                    break;
                }
            }
        }
        int size = mDateTimeList.size();
        int numOfDays = 42 - size;
        DateTime lastDateTime = mDateTimeList.get(size - 1);
        for (int i = 1; i <= numOfDays; i++) {
            WeakReference<DateTime> nextDateTime = new WeakReference<>(lastDateTime.plusDays(i));
            mDateTimeList.add(nextDateTime.get());
        }
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        long start = System.currentTimeMillis();
        this.mWidth = getWidth();
        this.mHeight = getHeight();
        if (mCells == null) {
            initCells();
        }
        for (int i = 0; i < ROWS * COLS; i++) {
            Rect rect = mCells.get(i);
            int color;
            DateTime dateTime = mDateTimeList.get(i);
            if (mYear != dateTime.getYear() || mMonth != dateTime.getMonth()) {
                color = Color.GRAY;
            } else {
                if (eventsMap.containsKey(dateTime)) {
                    drawEvents(canvas, eventsMap.get(dateTime), rect);
                }
                if (dateTime.getDay() == currentDay && dateTime.getMonth() == currentMonth && dateTime.getYear() == currentYear) {
                    color = mTodayColor;
                } else {
                    color = mDefaultColor;
                }
            }
            drawRectText(dateTime.getDay().toString(), canvas, rect, color);
        }
        LogUtil.d(TAG, "onDraw: " + (System.currentTimeMillis() - start));
    }

    private void drawEvents(Canvas canvas, Events events, Rect rect) {
        List<Rect> rects = circlesMap.get(rect);
        int index = 0;
        events.moveToStart();
        circlePaint.setAlpha(50);
        circlePaint.setStyle(Paint.Style.FILL);
        int maxEvents = GRID_R_C * GRID_R_C;
        while (events.hasNext() && index < maxEvents) {
            WeakReference<Events.Event> event = new WeakReference<>(events.getNext());
            circlePaint.setColor(event.get().getColor());
            Rect r = rects.get(index);
            int cX = r.centerX();
            int cY = r.centerY();
            if (index > 0 && index < maxEvents - 1) {
                WeakReference<Events.Event> prev = new WeakReference<>(events.getPreviousWithoutMoving());
                if (prev.get() != null) {
                    Rect end = rects.get(index - 1);
                    canvas.drawLine(cX, cY, end.centerX(), end.centerY(), circlePaint);
                }
            }
            canvas.drawCircle(r.centerX(), r.centerY(), r.width() / 4f, circlePaint);
            index++;
        }
    }

    private void initCells() {
        WeakReference<Rect> bounds = new WeakReference<>(new Rect());
        getLocalVisibleRect(bounds.get());
        int cellWidth = mWidth / COLS;
        int cellHeight = mHeight / ROWS;
        mCells = new ArrayList<>();
        for (int i = 0; i < ROWS; i++) {
            for (int j = 0; j < COLS; j++) {
                int top = i * cellHeight;
                int left = j * cellWidth;
                Rect tmp = new Rect(left, top, left + cellWidth, top + cellHeight);
                mCells.add(tmp);
                generateCircles(tmp);
            }
        }
    }

    private void generateCircles(Rect rect) {
        int circleWidth = rect.width() / GRID_R_C;
        int circleHeight = rect.height() / GRID_R_C;
        int rectTop = rect.top;
        int rectLeft = rect.left;
        List<Rect> rects = new ArrayList<>();
        for (int i = 0; i < GRID_R_C; i++) {
            for (int j = 0; j < GRID_R_C; j++) {
                int top = i * circleHeight + rectTop;
                int left = j * circleWidth + rectLeft;
                WeakReference<Rect> tmp = new WeakReference<>(new Rect(left, top, left + circleWidth, top + circleHeight));
                rects.add(tmp.get());
            }
        }
        circlesMap.put(rect, rects);
    }

    private void drawRectText(String text, Canvas canvas, Rect r, int color) {
        paint.setTextSize(30);
        paint.setTextAlign(Paint.Align.CENTER);
        paint.setAlpha(100);
        paint.setColor(color);
        int width = r.width();
        int numOfChars = paint.breakText(text, true, width, null);
        int start = (text.length() - numOfChars) / 2;
        canvas.drawText(text, start, start + numOfChars, r.exactCenterX(), r.exactCenterY(), paint);
    }

    @Override
    public boolean onTouch(View view, MotionEvent motionEvent) {
        switch (motionEvent.getAction()) {
            case MotionEvent.ACTION_DOWN:
                performTouch(motionEvent);
                break;
            case MotionEvent.ACTION_UP:
                performAction(motionEvent);
                break;
            case MotionEvent.ACTION_CANCEL:
                cancelTouch();
                break;
            case MotionEvent.ACTION_MOVE:
                performMove(motionEvent);
                break;
        }
        return true;
    }

    private void performAction(MotionEvent motionEvent) {
        int x = (int) motionEvent.getX();
        int y = (int) motionEvent.getY();
        mLongClickHandler.removeCallbacks(mLongRunnable);
        if (mTouchRect != null && mTouchRect.contains(x, y) && mDateClick != null) {
            mDateClick.onClick(mDateTimeList.get(mTouchPosition));
        }
        cancelTouch();
    }

    private void performMove(MotionEvent motionEvent) {
        int x = (int) motionEvent.getX();
        int y = (int) motionEvent.getY();
        if (mTouchRect != null && !mTouchRect.contains(x, y)) {
            cancelTouch();
        }
    }

    private void cancelTouch() {
        mTouchPosition = -1;
        mTouchRect = null;
    }

    private void performTouch(MotionEvent motionEvent) {
        int x = (int) motionEvent.getX();
        int y = (int) motionEvent.getY();
        for (int i = 0; i < ROWS * COLS; i++) {
            Rect rect = mCells.get(i);
            if (rect != null && rect.contains(x, y)) {
                mTouchPosition = i;
                mTouchRect = rect;
                mLongClickHandler.postDelayed(mLongRunnable, LONG_CLICK_TIME);
                break;
            }
        }
    }

    public interface OnDateClick {
        void onClick(DateTime dateTime);
    }

    public interface OnDateLongClick {
        void onLongClick(DateTime dateTime);
    }
}

File: app/src/main/java/com/elementary/tasks/core/views/MonthView.kt
package com.elementary.tasks.core.views

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.LinearGradient
import android.graphics.Paint
import android.graphics.Rect
import android.graphics.Shader
import android.graphics.Typeface
import android.os.Handler
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import androidx.annotation.IntRange
import androidx.core.content.ContextCompat
import androidx.core.content.res.ResourcesCompat
import com.elementary.tasks.R
import com.elementary.tasks.core.calendar.Events
import com.elementary.tasks.core.utils.Prefs
import com.elementary.tasks.core.utils.ThemeUtil
import com.elementary.tasks.core.utils.dp2px
import hirondelle.date4j.DateTime
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject
import timber.log.Timber
import java.lang.ref.WeakReference
import java.util.*

class MonthView : View, View.OnTouchListener, KoinComponent {

  private var mYear: Int = 0
  private var mMonth: Int = 0
  private var currentYear: Int = 0
  private var currentMonth: Int = 0
  private var currentDay: Int = 0
  private var mDateTimeList: MutableList<DateTime>? = null
  private var eventsMap: Map<DateTime, Events> = HashMap()

  private lateinit var paint: Paint
  private lateinit var circlePaint: Paint
  private lateinit var borderPaint: Paint
  private lateinit var touchPaint: Paint

  private lateinit var horizontalGradient: LinearGradient
  private lateinit var verticalGradient: LinearGradient
  private val gradientColors: IntArray by lazy {
    intArrayOf(
      Color.TRANSPARENT,
      ThemeUtil.colorWithAlpha(ThemeUtil.getThemeSecondaryColor(context), 90),
      Color.TRANSPARENT
    )
  }

  private var mCells: MutableList<Rect>? = null
  private val circlesMap = mutableMapOf<Rect, List<Rect>>()

  private var mWidth: Int = 0
  private var mHeight: Int = 0

  private var mDefaultColor: Int = 0
  private var mTodayColor: Int = 0

  private var mTouchPosition = -1
  private var mTouchRect: Rect? = null
  private var mNormalTypeface: Typeface? = null
  private var mBoldTypeface: Typeface? = null

  private val mLongClickHandler = Handler()
  private var mDateClick: OnDateClick? = null
  private var mDateLongClick: OnDateLongClick? = null
  private val mLongRunnable = object : Runnable {
    override fun run() {
      mLongClickHandler.removeCallbacks(this)
      if (mTouchRect != null && mDateLongClick != null) {
        mDateLongClick?.onLongClick(mDateTimeList!![mTouchPosition])
      }
      cancelTouch()
      invalidate()
    }
  }

  private val prefs: Prefs by inject()

  constructor(context: Context) : super(context) {
    init(context)
  }

  constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
    init(context)
  }

  constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
    init(context)
  }

  private fun init(context: Context) {
    val colorSecondary = ThemeUtil.getThemeSecondaryColor(context)

    this.mDefaultColor = ContextCompat.getColor(context, R.color.color_on_background)
    this.mTodayColor = ThemeUtil.themedColor(context, prefs.todayColor)

    this.mNormalTypeface = ResourcesCompat.getFont(context, R.font.roboto_regular)
    this.mBoldTypeface = ResourcesCompat.getFont(context, R.font.roboto_bold)

    this.borderPaint = Paint()
    this.borderPaint.style = Paint.Style.STROKE
    this.borderPaint.strokeWidth = dp2px(1).toFloat()

    this.touchPaint = Paint()
    this.touchPaint.isAntiAlias = true
    this.touchPaint.style = Paint.Style.FILL
    this.touchPaint.color = ThemeUtil.colorWithAlpha(colorSecondary, 50)

    this.paint = Paint()
    this.paint.isAntiAlias = true
    this.paint.typeface = mNormalTypeface

    this.circlePaint = Paint()
    this.circlePaint.isAntiAlias = true

    val calendar = Calendar.getInstance()
    calendar.timeInMillis = System.currentTimeMillis()
    this.currentDay = calendar.get(Calendar.DAY_OF_MONTH)
    this.currentMonth = calendar.get(Calendar.MONTH) + 1
    this.currentYear = calendar.get(Calendar.YEAR)
    setDate(calendar.get(Calendar.YEAR), calendar.get(Calendar.MONTH) + 1)
    setOnTouchListener(this)
  }

  fun setDateClick(dateClick: OnDateClick) {
    this.mDateClick = dateClick
  }

  fun setDateLongClick(dateLongClick: OnDateLongClick) {
    this.mDateLongClick = dateLongClick
  }

  fun setEventsMap(eventsMap: Map<DateTime, Events>) {
    this.eventsMap = eventsMap
    invalidate()
  }

  fun setDate(year: Int, @IntRange(from = 1, to = 12) month: Int) {
    mDateTimeList = ArrayList()
    mMonth = month
    mYear = year
    val firstDateOfMonth = DateTime(mYear, mMonth, 1, 0, 0, 0, 0)
    val lastDateOfMonth = firstDateOfMonth.plusDays(firstDateOfMonth.numDaysInMonth - 1)
    var weekdayOfFirstDate = firstDateOfMonth.weekDay
    val startDayOfWeek = prefs.startDay + 1
    if (weekdayOfFirstDate < startDayOfWeek) {
      weekdayOfFirstDate += 7
    }
    while (weekdayOfFirstDate > 0) {
      val dateTime = firstDateOfMonth.minusDays(weekdayOfFirstDate - startDayOfWeek)
      if (!dateTime.lt(firstDateOfMonth)) {
        break
      }
      mDateTimeList?.add(dateTime)
      weekdayOfFirstDate--
    }
    for (i in 0 until lastDateOfMonth.day) {
      mDateTimeList!!.add(firstDateOfMonth.plusDays(i))
    }
    var endDayOfWeek = startDayOfWeek - 1
    if (endDayOfWeek == 0) {
      endDayOfWeek = 7
    }
    if (lastDateOfMonth.weekDay != endDayOfWeek) {
      var i = 1
      while (true) {
        val nextDay = lastDateOfMonth.plusDays(i)
        mDateTimeList?.add(nextDay)
        i++
        if (nextDay.weekDay == endDayOfWeek) {
          break
        }
      }
    }
    val size = mDateTimeList?.size ?: 0
    val numOfDays = 42 - size
    val lastDateTime = mDateTimeList?.get(size - 1) ?: return
    for (i in 1..numOfDays) {
      val nextDateTime = WeakReference(lastDateTime.plusDays(i))
      mDateTimeList?.add(nextDateTime.get()!!)
    }
  }

  override fun onDraw(canvas: Canvas) {
    super.onDraw(canvas)
    val start = System.currentTimeMillis()
    this.mWidth = width
    this.mHeight = height
    if (mCells == null) {
      initCells()
    }
    for (i in 0 until ROWS * COLS) {
      val rect = mCells!![i]
      val dateTime = mDateTimeList!![i]
      var typeface = mNormalTypeface
      val color = if (mYear != dateTime.year || mMonth != dateTime.month) {
        Color.GRAY
      } else {
        if (eventsMap.containsKey(dateTime)) {
          val events = eventsMap[dateTime]
          if (events != null) {
            drawEvents(canvas, events, rect)
          }
        }
        if (dateTime.day == currentDay && dateTime.month == currentMonth && dateTime.year == currentYear) {
          typeface = mBoldTypeface
          mTodayColor
        } else {
          mDefaultColor
        }
      }
      drawRectText(dateTime.day.toString(), canvas, rect, color, i, typeface)
    }
    Timber.d("onDraw: ${(System.currentTimeMillis() - start)}")
  }

  private fun drawEvents(canvas: Canvas, events: Events, rect: Rect) {
    val rects = circlesMap[rect] ?: return
    var index = 0
    events.moveToStart()
    circlePaint.alpha = 50
    circlePaint.style = Paint.Style.FILL
    val maxEvents = GRID_R_C * GRID_R_C
    while (events.hasNext() && index < maxEvents) {
      val event = WeakReference(events.next)
      circlePaint.color = event.get()!!.color
      val r = rects[index]
      val cX = r.centerX()
      val cY = r.centerY()
      if (index > 0 && index < maxEvents - 1) {
        val prev = WeakReference<Events.Event>(events.previousWithoutMoving)
        if (prev.get() != null) {
          val end = rects[index - 1]
          canvas.drawLine(cX.toFloat(), cY.toFloat(), end.centerX().toFloat(), end.centerY().toFloat(), circlePaint)
        }
      }
      canvas.drawCircle(r.centerX().toFloat(), r.centerY().toFloat(), r.width() / 4f, circlePaint)
      index++
    }
  }

  private fun initCells() {
    val bounds = WeakReference(Rect())
    getLocalVisibleRect(bounds.get())
    val cellWidth = mWidth / COLS
    val cellHeight = mHeight / ROWS
    horizontalGradient = LinearGradient(0f, 0f, cellWidth.toFloat(), 0f,
      gradientColors, null, Shader.TileMode.MIRROR)
    verticalGradient = LinearGradient(0f, 0f, 0f, cellHeight.toFloat(),
      gradientColors, null, Shader.TileMode.MIRROR)
    mCells = ArrayList()
    for (i in 0 until ROWS) {
      for (j in 0 until COLS) {
        val top = i * cellHeight
        val left = j * cellWidth
        val tmp = Rect(left, top, left + cellWidth, top + cellHeight)
        mCells?.add(tmp)
        generateCircles(tmp)
      }
    }
  }

  private fun generateCircles(rect: Rect) {
    val circleWidth = rect.width() / GRID_R_C
    val circleHeight = rect.height() / GRID_R_C
    val rectTop = rect.top
    val rectLeft = rect.left
    val rects = ArrayList<Rect>()
    for (i in 0 until GRID_R_C) {
      for (j in 0 until GRID_R_C) {
        val top = i * circleHeight + rectTop
        val left = j * circleWidth + rectLeft
        val tmp = WeakReference(Rect(left, top, left + circleWidth, top + circleHeight))
        rects.add(tmp.get()!!)
      }
    }
    circlesMap[rect] = rects
  }

  private fun drawRectText(text: String, canvas: Canvas, r: Rect, color: Int, position: Int, typeface: Typeface?) {
    paint.textSize = dp2px(16).toFloat()
    paint.textAlign = Paint.Align.CENTER
    paint.alpha = 100
    paint.color = color
    paint.style = Paint.Style.FILL_AND_STROKE
    paint.typeface = typeface

    if (position == mTouchPosition) {
      canvas.drawRect(r, touchPaint)
    }

    val width = r.width()
    val numOfChars = paint.breakText(text, true, width.toFloat(), null)
    val start = (text.length - numOfChars) / 2
    canvas.drawText(text, start, start + numOfChars, r.exactCenterX(), r.exactCenterY(), paint)
    if (position == 0 || ((position - 6) % 7) != 0) {
      borderPaint.shader = verticalGradient
      canvas.drawLine(r.right.toFloat(), r.top.toFloat(),
        r.right.toFloat(), r.bottom.toFloat(), borderPaint)
    }
    if (position <= 34) {
      borderPaint.shader = horizontalGradient
      canvas.drawLine(r.left.toFloat(), r.bottom.toFloat(),
        r.right.toFloat(), r.bottom.toFloat(), borderPaint)
    }
  }

  override fun onTouch(view: View, motionEvent: MotionEvent?): Boolean {
    if (motionEvent == null) return true
    when (motionEvent.action) {
      MotionEvent.ACTION_DOWN -> performTouch(motionEvent)
      MotionEvent.ACTION_UP -> performAction(motionEvent)
      MotionEvent.ACTION_CANCEL -> cancelTouch()
      MotionEvent.ACTION_MOVE -> performMove(motionEvent)
    }
    return true
  }

  private fun performAction(motionEvent: MotionEvent) {
    val x = motionEvent.x.toInt()
    val y = motionEvent.y.toInt()
    mLongClickHandler.removeCallbacks(mLongRunnable)
    if (mTouchRect != null && mTouchRect?.contains(x, y) == true && mDateClick != null) {
      mDateClick?.onClick(mDateTimeList!![mTouchPosition])
    }
    cancelTouch()
    invalidate()
  }

  private fun performMove(motionEvent: MotionEvent) {
    val x = motionEvent.x.toInt()
    val y = motionEvent.y.toInt()
    if (mTouchRect != null && mTouchRect?.contains(x, y) == false) {
      cancelTouch()
      invalidate()
    }
  }

  private fun cancelTouch() {
    mTouchPosition = -1
    mTouchRect = null
  }

  private fun performTouch(motionEvent: MotionEvent) {
    val x = motionEvent.x.toInt()
    val y = motionEvent.y.toInt()
    for (i in 0 until ROWS * COLS) {
      val rect = mCells!![i]
      if (rect.contains(x, y)) {
        mTouchPosition = i
        mTouchRect = rect
        mLongClickHandler.postDelayed(mLongRunnable, LONG_CLICK_TIME)
        invalidate()
        break
      }
    }
  }

  interface OnDateClick {
    fun onClick(dateTime: DateTime)
  }

  interface OnDateLongClick {
    fun onLongClick(dateTime: DateTime)
  }

  companion object {
    private const val ROWS = 6
    private const val COLS = 7
    private const val GRID_R_C = 3
    private const val LONG_CLICK_TIME: Long = 500
  }
}
